.grit/patterns/moment-to-date-fns.grit

```diff
@@ -0,0 +1,765 @@
+      $return = `week`
+    } else if ($specifier <: regex_day()) {
+      $return = `day`
+    } else if ($specifier <: regex_hour()) {
+      $return = `hour`
+    } else if ($specifier <: regex_minute()) {
+      $return = `minute`
+    } else if ($specifier <: regex_second()) {
+      $return = `second`
+    } else if ($specifier <: regex_ms()) {
+      $return = `millisecond`
+    },
+    $norm = $return
+  }
+}
+
+// add/subtract: convert all momentjs arithmetic operations to date-fns operations.
+// eg: today.add(oneday) => add(today, oneday)
+pattern rewrite_moment_arithmetic() {
+  or {
+    // add/subtract: convert all momentjs arithmetic operations to date-fns operations.
+    // eg: today.add(oneday) => add(today, oneday)
+    moment_arith($date, $method, $count, $specifier) as $expression where {
+      $specifier <: and {
+        string(),
+        or {
+          year() => `years`,
+          quarter() => `quarters`,
+          month() => `months`,
+          week() => `weeks`,
+          day() => `days`,
+          hour() => `hours`,
+          minute() => `minutes`,
+          second() => `seconds`,
+          millisecond() => `milliseconds`
+        }
+      },
+
+      $out = `$method($date, { $specifier: $count })`,
+
+      if ($date <: identifier()) {
+        $out = `$date = $method($date, { $specifier: $count })`
+      },
+      // If the arithmetic expression is not an expression statement,
+      // wrap it in `()`, as `=` has very low precedence.
+      if ($date <: not within expression_statement($expression)) {
+        $out = `($out)`
+      }
+    } => `$out`,
+
+    moment_arith($date, $method, $count, $specifier) where {
+      !$specifier <: string(),
+      $out = `$method($date, { [normalizeMomentJSUnit($specifier) + 's']: $count })`,
+      if ($date <: identifier()) { $out = `$date = $out`},
+      if ($date <: not within expression_statement($expression)) {
+        $out = `($out)`
+      }
+    } => `$out`,
+
+    `$date_a.diff($date_b, $literal_unit)` where { $literal_unit <: string() }  => `datefns.differenceIn$literal_unit($date_a, $date_b)` where {
+      $literal_unit <: or {
+        year() => `Years`,
+        quarter() => `Quarters`,
+        month() => `Months`,
+        week() => `Weeks`,
+        day() => `Days`,
+        hour() => `Hours`,
+        minute() => `Minutes`,
+        second() => `Seconds`,
+        millisecond() => `Milliseconds`
+      }
+    },
+
+    // When no unit is specified, `diff` returns difference in ms.
+    `$date_a.diff($date_b)` where { $date_b <: [$_] /* exactly one argument */ } => `datefns.differenceInMilliseconds($date_a, $date_b)`,
+  }
+}
+
+pattern rewrite_moment_getters_and_setters() {
+  or {
+    `$date.daysInMonth()` => `datefns.getDaysInMonth($date)`,
+     moment_getter_or_setter(),
+
+    `$date.$func($arg)` as $exp where {
+      $func <: or {
+        r"dates?" => `Month`,
+        r"(?:days?)|(?:weekdays?)" => `Day`,
```

Comment:
Avoid regex when feasible.

```suggestion
       or { `date`, `dates` } => `Month`,
       or { `day`, `days`, `weekday`, `weekdays` } => `Day`,
```

(In theory we can add a compiler optimization for this. But for now this is better.)